### PR TITLE
publish to npm directly from electron repo 

### PR DESCRIPTION
**Pull request by vanessayuenn**
This is an effort to allow publishing electron to npm directly from the main repo, with the goal of eventually decommissioning the [electron-prebuilt-updater heroku app](https://github.com/electron/electron-prebuilt-updater). The vast majority of the code is ported over from the existing [electron-prebuilt repo](https://github.com/electron-userland/electron-prebuilt).

### CODE REVIEW

File: script/publish-to-npm.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      env: Object.assign({}, process.env, { electron_config_cache: tempDir }),
+      cwd: tempDir
+    })
+    const checkVersion = childProcess.execSync(`${path.join(tempDir, 'node_modules', '.bin', 'electron')} -v`)
+    assert.strictEqual(checkVersion.toString().trim(), `v${rootPackageJson.version}`)
+    resolve(tarballPath)
+  })
+})
+.then((tarballPath) => childProcess.execSync(`npm publish ${tarballPath} --tag ${npmTag}`))
+.catch((err) => console.error(`Error: ${err}`))
```
Comment (kevinsawicki) [inline script/publish-to-npm.js]: Might want to `process.exit(1)` after logging the error so the script exits unsuccessfully

File: script/publish-to-npm.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    request.get({
+      url: tsdAsset.url,
+      headers: {
+        'accept': 'application/octet-stream',
+        'user-agent': 'electron-npm-publisher'
+      }
+    }, (err, response, body) => {
+      if (err) {
+        reject(err)
+      } else {
```
Comment (kevinsawicki) [inline script/publish-to-npm.js]: Might be worth checking for a `200` status here to ensure that something like a `404` page isn't written to `electron.d.ts`

File: script/publish-to-npm.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    })
+  })
+})
+.then((release) => {
+  npmTag = release.prerelease ? 'beta' : 'latest'
+})
+.then(() => childProcess.execSync('npm pack', { cwd: tempDir }))
+.then(() => {
+  // test that the package can install electron prebuilt from github release
+  const tarballPath = path.join(tempDir, `electron-${rootPackageJson.version}.tgz`)
```
Comment (kevinsawicki) [inline script/publish-to-npm.js]: Maybe use `${rootPackageJson.name}-${rootPackageJson.version}.tgz` to make it more generic/configurable?

File: script/upload.py [hunk 4]
```diff
@@ -72,6 +72,8 @@ def main():
     # Press the publish button.
     publish_release(github, release['id'])
 
+    # TODO: run publish-to-npm script here
```
Comment (zeke) [inline script/upload.py]: @vanessayuenn do you think we should leave this as-is and publish manually for now? Or would you rather implement it before we ship?

File: npm/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
+# electron-prebuilt
+
+[![Travis build status](http://img.shields.io/travis/electron-userland/electron-prebuilt.svg?style=flat)](http://travis-ci.org/electron-userland/electron-prebuilt)
```
Comment (zeke) [inline npm/README.md]: These badges won't be accurate any more. Should we point them at electron/electron? Take them out?

File: script/upload.py (hunk 4 above)
Comment (vanessayuenn) [inline script/upload.py]: The plan is to keep this as-is and run the script manually for now. The comment here is just a note for when we're ready to chain it into the publishing process in the future.

File: npm/README.md (hunk 5 above)
Comment (vanessayuenn) [inline npm/README.md]: Maybe we can just take this README out completely, and use the one from root? That way we can keep the docs on the npm package page consistent with what we have up on the github repo.

Comment (zeke) [inline npm/README.md]: > use the one from root?

I like the idea of npm.im/electron having the same readme as ghub.io/electron, as that is what users normally expect. But if we do that, I think we'll need to preserve some parts of the prebuilt readme, as it has useful installation and troubleshooting info that's not present in the top-level readme.

**Comment by zeke** [comment]
> Could we possibly squash these commits down

@MarshallOfSound I think the intent here was to preserve the history so folks are recognized for their contributions.

I personally am indifferent as to whether we preserve the commits or squash them.

**Comment by vanessayuenn** [comment]
Hey @MarshallOfSound, just echoing what @zeke was saying, the commits history was ported over on purpose to preserve the contributions made in the past. I think if the number of commits proves to cause issues upon merging the PR, we can then resort to squashing.

**Pull request merged**